### ISSUE TITLE
Modification to atari ST config files to get working disks

### ISSUE THREAD

**Issue by stevencorby** [davidgiven/fluxengine#337]
I have some Atari ST disk images in the form of raw .st files.  I used Fluxengine to write them to real disks, but was unable to get the ST to read any of them successfully.  The machine would either hang on bootup or sit forever with a bee icon after opening the drive in the desktop.

As an experiment I tried the ibm720 config on one of the atari .st 720K images and it worked perfectly.

I did some tests to try and figure out what was different about the IBM and Atari config files that yields a working disk.  In the end I discovered that the sector numbering seems to be the factor that fixed the problem.

The IBM config image_reader start_sector is 1, whilst the atarist config start_sector is 0.  In addition, the encoder sectors number from 1-9 for the IBM, but for the Atari they are numbered 0-8/9.    

By changing the Atari config start_sector to 1 instead of 0, and the encoder sectors to 1-9 (or 1-10 for 10 sector/track images), I was able to get my ST to load the disks without any problems.

I don't pretend to understand what I've done here - and I haven't looked at the source code, so I can't offer any explanation as to what's going on.

Is it simply the case that the config files are wrong for the Atari - or do my results indicate that there may be genuine bug? When I get chance I'll dig further to try and understand what's going on and provide an explanation (unless someone can beat me to it).

Thanks

**Comment by stevencorby** [comment]
Hi David,

I’ve tested the Atari branch with a 720K, 810K and 820K image and they all seem to write fine.  (Note I used the atarist820 config for both the 820K and 810K image – I presume it just writes an empty 82nd track).   I also tried reading an 820K cover disk I had kicking around and then writing it back to a blank disk.  That seems to work fine too.

Just to satisfy my curiosity can you summarise the significance of changing the start sector value from 0 to 1? Is this value supposed to start at 0 or 1?  I’ve got another old Atari 800 with a drive that I’m planning to restore and I intend to build a 5 ¼ version of fluxengine and write my own config file (which I will send you assuming I get it to work).  But understanding what’s going on here would help.

If you want me to test any changes to fluxengine in the future with my Atari ST then just drop me an email and I’ll be happy to report back.

Many thanks, Steve

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by davidgiven** [comment]
Excellent --- thanks for testing; I'll merge that.

So the issue with sector IDs is that they're actually written onto the disk as part of the sector header, and can be completely arbitrary. Various platforms use various numbering schemes. The most common (e.g. IBM PCs) is 1-based but some (e.g. Acorn machines) use 0-based. The Atari uses 1-based, like PCs, but I broke it in a recent refactor.

BTW, if you know about Atari STs... the one I've got has a faulty 3.5" drive, but the Atari drive selection logic is different from normal PC drives so I can't make FluxEngine talk to it. I want to fix this; do you know where I can find a reference of how the Atari drives differ from PC drives?

**Comment by stevencorby** [comment]
Thanks for explaining the sector ID puzzle for me - it's satisfying to understand what it was I was changing,

WRT ST hardware, I'm certainly no expert I'm afraid.  But I do have the luxury of owning a working ST drive.  I tried hooking the ST drive up to FluxEngine and it worked perfectly.  I was able to write a disk with FluxEngine using the ST drive and then read it back on the ST.  The only thing I had to do was change the ST drive select from 0 to 1 (ST expects 0, FluxEngine expects 1).  Fortunately my drive has a header that made this easy.

There's some info here on how to get a PC drive to work in an ST [https://www.atari-forum.com/viewtopic.php?t=14049](url)  I personally haven't tried it.  Not sure I could provide any detail that isn't already in the forums - but if you have any experiments you want me to try I'm happy to oblige.

Thanks again.
